File: demos_code_J1_conception_objet/04_couplage_paiement/DemoCouplageAvant.java
/**
 * Simulation locale d'un SDK externe : aucune vraie dépendance réseau.
 */
public class DemoCouplageAvant {
    static final class StripeResponse {
        private final String status;
        StripeResponse(String status) { this.status = status; }
        String status() { return status; }
    }

    static final class StripeClient {
        StripeResponse charge(int cents) {
            System.out.println("[Stripe SDK simulé] charge " + cents + " cents");
            return new StripeResponse("paid");
        }
    }

    static final class ValiderCommande {
        private final StripeClient stripe = new StripeClient();

        boolean executer(int totalCents) {
            StripeResponse response = stripe.charge(totalCents);
            return "paid".equals(response.status());
        }
    }

    public static void main(String[] args) {
        boolean ok = new ValiderCommande().executer(4200);
        System.out.println("Paiement accepté = " + ok);
    }
}

File: Demo en Python/demos_code_J1_conception_objet_python/04_couplage_paiement/DemoCouplageAvant.py
from dataclasses import dataclass


@dataclass(frozen=True)
class StripeResponse:
    status: str


class StripeClient:
    """Simulation locale d'un SDK externe : aucune vraie dépendance réseau."""

    def charge(self, cents: int) -> StripeResponse:
        print(f"[Stripe SDK simulé] charge {cents} cents")
        return StripeResponse("paid")


class ValiderCommande:
    def __init__(self) -> None:
        # Dépendance créée en dur : le cas d'usage connaît Stripe.
        self.stripe = StripeClient()

    def executer(self, total_cents: int) -> bool:
        response = self.stripe.charge(total_cents)
        return response.status == "paid"


def main() -> None:
    ok = ValiderCommande().executer(4200)
    print(f"Paiement accepté = {ok}")


if __name__ == "__main__":
    main()
